### PR TITLE
Improve rest api

### PR DESCRIPTION
**Pull request by flawedmatrix**
Tons of changes were made to the website interface:
- Changed all (I hope) instances of urls to use reverse() for looking up the API
- Accessing the list of games is now GET /game/
- Accessing a game with id {gameid} is now GET /game/{gameid}
- Accessing a map with id {mapid} is now GET /map/{mapid}
- Saving a map to {mapid} is now POST /map/{mapid}
- Saving a new map without a {mapid} is now POST /map/
- gamens.py is now sockets.py (if gevent-socketio gets updated, this will be compatible with the future version)
- game.views and map.views now contain class-based views, which makes it easier to separate GET and POST requests and handle edge cases
- Moved Game controller methods to models.py, which makes them easier to test
- Removed all of the login_required decorators, since it was very repetitive
- Added LoginRequiredMiddleware, which basically puts login_required on every request except the ones specified in settings.LOGIN_EXEMPT_URLS (which can handle regexps) and settings.LOGIN_URL
- Tests were modified to support the new API
- Changed the jasmine tests to use the new api (although the tests correctly use the api, it is unable to access the website because the new login required middleware redirects it back to the homepage)
- Aborted writing tests for the new API... the map URLs were already tested pretty thoroughly as far as I know and the game views are hard to test.

### CODE REVIEW

**Comment by hellochar** [comment]
I think the routes are to `map/mapid` instead of `map/#mapid` and the like, right?

at least, i think they should be that way

File: the_game_bazaar/templates/the_game_bazaar/play.html [hunk 1]
```diff
@@ -17,12 +17,66 @@
                 <div id="play-join">
                     <button id="play-host-button" class="btn btn-primary pull-right">Host a Game</button>
                     <h3>Join A Game:</h3>
-                    <div id="play-join-table"></div>
+                    <div id="play-join-table">
```
Comment (hellochar) [inline the_game_bazaar/templates/the_game_bazaar/play.html:20]: Why did we change populating this table dynamically through javascript back into embedding it into the template? Don't we still need dynamic loading later on for when we want to refresh the games list automatically?

File: the_game_bazaar/urls.py [hunk 2]
```diff
@@ -1,23 +1,27 @@
...
+lib.sdjango.autodiscover()
+
+urlpatterns = patterns(
+    '',
     url(r'^$',
         views.index,
         name="index"),
 
-    url(r'^socket.io/', include('game.urls')),
+    url(r'^socket\.io', include(lib.sdjango.urls)),
```
Comment (hellochar) [inline the_game_bazaar/urls.py:19]: LOL good catch on the regex

Comment (hellochar) [inline the_game_bazaar/urls.py:19]: Also does having a trailing slash in the regex change the behavior of the routes?

File: game/forms.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
+# forms.py
+from django import forms
+
+
+class GameForm(forms.Form):
+    map_id = forms.IntegerField(required=True, min_value=0)
```
Comment (hellochar) [inline game/forms.py:6]: django forms look really useful; perhaps we should use them for more ajax calls and etc.

File: game/static/game/js/game.js [hunk 4]
```diff
@@ -178,13 +178,7 @@ Game.prototype.instantiateGameState = function() {
...
     $.ajax({
         type: "GET",
-        url: "/map/",
-        data: {
-            "map_id": parseInt(this.map_id, 10)
-        },
-        headers: {
-            "X-CSRFToken": $.cookie('csrftoken')
-        },
+        url: "/map/" + parseInt(this.map_id, 10),
```
Comment (hellochar) [inline game/static/game/js/game.js:181]: parseInt doesn't do what you think it does (and I think it saw it in other places too). 

`The parseInt() function parses a string and returns an integer.`

This code will work because JS tries to blur the distinction between Numbers and Strings but in general we shouldn't need parseInt at all

File: editor/static/editor/js/editor.js [hunk 5]
```diff
@@ -48,10 +48,14 @@ Editor.prototype.setEditingMap = function(map) {
 
 Editor.prototype.saveMap = function() {
     var self = this;
-    data = {map_id : this.map.id, map_data : JSON.stringify(this.map.toJSON())};
+    var map_id = this.map.id;
```
Comment (hellochar) [inline editor/static/editor/js/editor.js:51]: where is this variable used?

**Comment by hellochar** [comment]
ohh i think you're confused about what `#mapid` means; in a url, the part after a # is called an anchor and has a special meaning. It's only accessible client-side. It is not some notation to mean that `mapid` is a number or anything. I use it in `/editor/#mapid` because you can dynamically change which map you are editing without reloading the page, and having an anchor helps people share the URL.

File: the_game_bazaar/templates/the_game_bazaar/play.html (hunk 1 above)
Comment (flawedmatrix) [inline the_game_bazaar/templates/the_game_bazaar/play.html:20]: I think for now we are fine with embedding this through the template, but for in the future when we do a dynamically updating table I left the ajax API for it in the urls.py.

Edit: Also the javascript for this was pretty ugly and it could have been simplified by having this table structure already preloaded.

File: game/static/game/js/game.js (hunk 4 above)
Comment (flawedmatrix) [inline game/static/game/js/game.js:181]: I had no clue what it was doing originally but it seemed to work so I just left it as that.

**Comment by flawedmatrix** [comment]
I understand that the # is meant to be used as an anchor but I wanted some way of indicating that it's not actually "mapid" in the URL.

I should probably specify a different way of representing what I meant by having a separate URL per gameid or mapid.

 Also to add, I wasn't planning on making /editor/#mapid a page loaded every time you want to change which map you are editing since it seems it would make more sense to edit different maps on the same page.

File: editor/static/editor/js/editor.js (hunk 5 above)
Comment (flawedmatrix) [inline editor/static/editor/js/editor.js:51]: Good catch! I wanted to use this variable but I forgot to change the code underneath.

File: the_game_bazaar/urls.py (hunk 2 above)
Comment (flawedmatrix) [inline the_game_bazaar/urls.py:19]: Forgot to respond to this one...

For some reason the regexes seem to be matching correctly with or without the trailing slashes. I guess we should have a specific policy about this in any case.

**Comment by hellochar** [comment]
ok in general it looks good; please change the documentation to not have the `#` in the hashtag because people will misinterpret it as being an anchor
